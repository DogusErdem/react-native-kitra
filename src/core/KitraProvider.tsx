import type { ComponentType, ForwardedRef } from 'react';
import React from 'react';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import Animated from 'react-native-reanimated';
import { StyleProp, ViewStyle } from 'react-native';

import { ThemeProvider } from './theme/theme';
import { TypographyContext, TypographyProvider } from './typography/typography';
import { NotificationContextType, NotificationProvider } from './notification/notification';

const Feather = require('react-native-vector-icons/Feather').default;
const Ionicons = require('react-native-vector-icons/Ionicons').default;
const AntDesign = require('react-native-vector-icons/AntDesign').default;

Animated.addWhitelistedNativeProps({ text: true });
const messageTypes = (theme:any) => ({
  SUCCESS: {
    backgroundColor: theme.colors.status.successLight,
    icon: <Feather name="check" size={25} color={theme.colors.status.success} />,
  },
  INFO: {
    backgroundColor: theme.colors.system.primary15,
    icon: <Ionicons name="information-circle-outline" size={25} color={theme.colors.system.primary} />,
  },
  DANGER: {
    backgroundColor: theme.colors.status.errorLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.error} />,
  },
  WARNING: {
    backgroundColor: theme.colors.status.warningLight,
    icon: <AntDesign name="warning" size={25} color={theme.colors.status.warning} />,
  },
});

type KitraProviderType= {
  children?:any,
  limit?:number,
  notificationRef?:React.Ref<NotificationContextType>,
  messageType?:(theme?:any, colorScheme?:'dark' | 'light') => {[key:string]: {
      backgroundColor:string,
    icon?:React.ReactNode
    onPress?:()=>void
  }},
  notificationCcontainerStyle?:StyleProp<ViewStyle>,
  customView?:({ header, type, message, theme }:
    {header:string, type:string, message:string, theme:any})=>React.ReactNode}

export const KitraProvider =
({ children, messageType = messageTypes, notificationCcontainerStyle, limit, customView,
  notificationRef }: KitraProviderType) => (
    <GestureHandlerRootView style={{ flex: 1 }}>
      <ThemeProvider>
        <TypographyProvider>
          <NotificationProvider
            messageType={theme => (messageType ? messageType(theme) : theme)}
            notificationCcontainerStyle={notificationCcontainerStyle}
            limit={limit}
            customView={item => customView?.(item)}
            ref={notificationRef}
          >
            {children}
          </NotificationProvider>
        </TypographyProvider>
      </ThemeProvider>
    </GestureHandlerRootView>
);

export function applyDefaults<T>(Component: ComponentType<T>) {
  return React.forwardRef((props: T, ref: ForwardedRef<any>) => (
    <TypographyContext.Consumer>
      {typography =>
        <Component ref={ref} typography={typography?.typography} {...props} />
          }
    </TypographyContext.Consumer>
  ));
}
