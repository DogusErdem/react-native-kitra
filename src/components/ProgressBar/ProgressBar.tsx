import { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import Animated, { useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated';
import useComponentTheme from '../../core/hooks/useComponentTheme';
import type { FCCWD, ProgressBarProps } from '../../types';
import { applyDefaults } from '../../core/KitraProvider';

const ProgressBar: FCCWD<ProgressBarProps> = (
  { progress = 0,
    progressStyle,
    barStyle,
    testID,
    theme },
) => {
  const { componentTheme } = useComponentTheme(theme, 'progressBar');
  const offset = useSharedValue(progress);
  const [barWidth, setBarWidth] = useState(0);
  offset.value = progress;
  const animatedStyle = useAnimatedStyle(() => ({
    width: withTiming((offset.value / 100) * barWidth, {
      duration: 400,
    }),
  }));

  return (
    <View
      testID={testID}
      onLayout={x => setBarWidth(x.nativeEvent.layout.width)}
      style={[Style.container, barStyle, { backgroundColor: componentTheme.default?.bar }]}
    >
      <Animated.View
        testID="progressBarAnimation"
        style={
          [
            animatedStyle,
            Style.progressStyle,
            progressStyle,
            {
              backgroundColor: componentTheme.default?.progress,
            },
          ]
        }
      />
    </View>

  );
};
export default applyDefaults(ProgressBar);

export const Style = StyleSheet.create({
  container: { backgroundColor: 'transparent', height: 2, width: '100%', overflow: 'visible', justifyContent: 'center' },
  progressStyle: { borderRadius: 14, position: 'absolute', height: 4 },
});
