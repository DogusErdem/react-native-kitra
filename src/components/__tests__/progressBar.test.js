import { fireEvent, render, act } from '@testing-library/react-native';
import { Text } from 'react-native';
import { KitraProvider } from '../../core/KitraProvider';
import ProgressBar from '../ProgressBar/ProgressBar';

/* eslint-disable no-undef */
const renderWithContext = children => render(<KitraProvider>{children}</KitraProvider>);

describe('ProgressBar', () => {
  it('default render', () => {
    const tree = renderWithContext(<ProgressBar />);
    expect(tree).toMatchSnapshot();
  });
  describe('render with custom styles', () => {
    it('render with progressStyle and barStyle props', () => {
      const { getByTestId } = renderWithContext(<ProgressBar
        testID="progressBar"
        barStyle={{ margin: 10, padding: 10, marginTop: 12 }}
        progressStyle={{ margin: 5, padding: 15, marginTop: 11 }}
      />);
      const bar = getByTestId('progressBar');
      const progress = getByTestId('progressBarAnimation');
      expect(bar).toHaveStyle({ margin: 10, padding: 10, marginTop: 12 });
      expect(progress).toHaveStyle({ margin: 5, padding: 15, marginTop: 11 });
    });
  });
  describe('render with custom theme', () => {
    it('render with default theme', () => {
      const { getByTestId } = renderWithContext(<ProgressBar
        testID="progressBar"
        theme={{ default: { bar: 'red', progress: 'white' } }}
      />);
      const bar = getByTestId('progressBar');
      const progress = getByTestId('progressBarAnimation');
      expect(bar).toHaveStyle({ backgroundColor: 'red' });
      expect(progress).toHaveStyle({ backgroundColor: 'white' });
    });
  });
});

// ANİMASYON TESTİ EKSİK
