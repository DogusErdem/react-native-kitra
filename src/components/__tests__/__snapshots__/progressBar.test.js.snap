// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar default render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "height": 2,
          "justifyContent": "center",
          "overflow": "visible",
          "width": "100%",
        },
        undefined,
        {
          "backgroundColor": "#E6EEFF",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "width": 0,
          },
          {
            "borderRadius": 14,
            "height": 4,
            "position": "absolute",
          },
          undefined,
          {
            "backgroundColor": "#195CEF",
          },
        ]
      }
      testID="progressBarAnimation"
    />
  </View>
</View>
`;
